### PR TITLE
Narrow array on count(), string on strlen() with positive-int

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/11558

### CODE REVIEW

File: src/Analyser/TypeSpecifier.php [hunk 1]
```diff
@@ -1096,87 +1101,6 @@ private function specifyTypesForConstantBinaryExpression(
 			));
 		}
 
-		if (
```
Comment (staabm) [inline src/Analyser/TypeSpecifier.php:1099]: moved this 2 if-blocks out of `specifyTypesForConstantBinaryExpression` as we no longer check for a constant type operand, but do a more general `isInteger` based handling, which includes unbound `IntegerRangeType`.

File: src/Analyser/TypeSpecifier.php [hunk 2]
```diff
@@ -1037,7 +1038,7 @@ private function turnListIntoConstantArray(FuncCall $countFuncCall, Type $type,
 					$offsetType = new ConstantIntegerType($i);
 					$valueTypesBuilder->setOffsetValueType($offsetType, $type->getOffsetValueType($offsetType), true);
 				}
-			} else {
+			} elseif ($type->isConstantArray()->yes()) {
```
Comment (staabm) [inline src/Analyser/TypeSpecifier.php:1041]: compensated the now passed in maybe unbound `IntegerRangeType`s

File: tests/PHPStan/Analyser/nsrt/count-type.php [hunk 3]
```diff
@@ -44,12 +44,12 @@ public function doFooBar(
 		if (count($arr) == $maybeZero) {
 			assertType('array', $arr);
 		} else {
-			assertType('non-empty-array', $arr);
+			assertType('array', $arr);
```
Comment (staabm) [inline tests/PHPStan/Analyser/nsrt/count-type.php:47]: I think I did a mistake here in previous PR, as the array can still be empty in the else branch

https://3v4l.org/409Jp

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Too many changes. I can't read that diff. Please send this in smaller chunks.

**Pull request closed without merging**